### ISSUE TITLE
Queue stuck with jobs in "active" state for hours, and rest of jobs not being processed

### ISSUE THREAD

**Issue by adamreisnz** [taskforcesh/bullmq#652]
Hello, we are finding that one of our queues is periodically getting stuck, with 2 jobs sitting in "active" and not being progressed for 18 hours, while the "waiting" queue is filling up with 1000's of jobs as a result.

The jobs failed with an error, but are not being removed from the "active" state. What can be done to address this? The queue should not get stuck as a result requiring manual intervention, as this is making bullmq completely unusable for us.

Please let me know what further information you need to help analyse this. Please don't ask for a "script to reproduce" because we don't have one. It's a complicated app with dozens of different jobs, and the jobs that are failing are not always failing, just occasionally when Chrome fails to load on the instance.

Thanks.

**Comment by manast** [comment]
With the information you provide if I have to guess I would say that your workers are not propagating the exceptions properly so that failure you are mentioning is leaving some promise hanging forever. Make sure that all your async chain is properly built and that all errors are thrown properly and promises returned.

**Comment by adamreisnz** [comment]
Ok, I will look into that from our end, although everything is wrapped in a neat job runner class with proper error handling and propagation, and all promises are run using async/await. Nevertheless, I will investigate when I'm back from holiday.

In the meantime, is there no timeout option for long running/hanging jobs? I would say after 18 hours it's pretty clear a job isn't going to finish.

Can we configure a timeout if this isn't done by default?

**Comment by manast** [comment]
We have not ported TTL yet from Bullv3, however in your case it would just hide a possible bug. If the jobs are active and there are workers working on them then it is because the job has not yet completed or failed. Since you mentioned you got an error from the worker the only explanation is that the exception is not thrown up all the way to the processor call.

**Comment by sven-codeculture** [comment]
Since updating bullmq from 1.14.x to 1.39.x we are having the same issue. I even added a timeout attribute to our jobs by default (60000ms timeout), but even a simple job that just fetches some data via curl (axios library) and should create addition jobs just keeps stuck for hours. (Jobs didn't change and before it was working like a charm)

The timeout i've set is never executed.
<img width="548" alt="Bildschirmfoto 2021-07-22 um 08 53 04" src="https://user-images.githubusercontent.com/3930643/126600126-76b35cd2-6f8d-4c6b-b45b-a32e61564062.png">

At time of writing this was stuck for two hours. Only a restart of the worker helps temporarily solving the problem.

**Comment by manast** [comment]
if you create a worker that doesn't do anything, just returns a resolved promise, what happens then?

**Comment by manast** [comment]
Btw, why don't downgrade if an older version worked perfectly for you?

**Comment by sven-codeculture** [comment]
The jobs get randomly stuck, most of the time they are running though as expected. Sometimes they randomly get stuck.
Its also not the same job type. Sometimes its a call to the API that gets stuck, sometimes a simple job that should add some data to the database.
I cannot downgrade again since i updated due to an even worse ticket. When the redis Server in our AWS ElastiCache did a fail over at the weekend the workers just stopped working. This should have been caused by the loop throwing an exception but from outside noone was able to find out that the loop stopped. Thats why i added https://github.com/taskforcesh/bullmq/pull/637 and after it was merged i updated to the new version.

@manast do you have any idea why the timeout is not doing anything? It would be ok for us it those jobs just die after 10 minutes and get retried

**Comment by manast** [comment]
@sven-codeculture As I said before, the TTL has not been ported from Bullv3 yet, mostly because we do not have promise cancel support so a better mechanism needs to be developed.

When you say "stucked" it does not help me so much in understanding the problem. Can you be more precise?

**Comment by sven-codeculture** [comment]
@manast we currently got a infrastructure with 3 workers running, all on a concurrency of 4. So we are able to process 12 Jobs at the same time.
We got different types of Jobs running. Some call the REST API for new data and create new jobs. Those jobs are then adding the data to our database using an GraphQL endpoint. Some jobs are just for cleaning up data using GraphQL of generating statistics for the customer.
There are about 20-100k jobs depending on how active the customers are that are run per day currently. Of those 2-3 per day get "stuck" meaning a job that normally takes at max 30s to complete is still sitting in "active" state after 12 hours without doing anything anymore. It already happend to jobs just doing a axios API call, and also jobs that are calling the GraphQL API to add or fetch data from there.

**Comment by manast** [comment]
@sven-codeculture but do you know if the job did everything it supposed to do? did it succeed in making the Axios call?, can you put logs precisely before the job completes to check if the job is indeed completing or if it is just hanging in some dangling promise?

**Comment by sven-codeculture** [comment]
Ill try to check this by extending the logging. However this only happened after the upgrade from 1.14 to 1.39, so i would be wondering if it's our jobs fault.

**Comment by manast** [comment]
@sven-codeculture It is not a matter of who's the fault, but to pinpoint the most exact place where it could be. For me it is impossible to find a bug in such a large range of changes specially when I cannot reproduce it myself...

**Comment by adamreisnz** [comment]
Well there's at least two instances being reported here of a seemingly wide
spread bug which causes random jobs to get stuck.

Seems to me it at least requires some further investigation.

I will report back later to see what I can find in our tasks.

However, it would be good if the TTL functionality can be implemented as
soon as possible so that we can at least avoid jobs running indefinitely,
without having to build this functionality ourselves.

**Comment by manast** [comment]
@adamreisnz I am not sure "wide spread bug" is the proper definition, since it seems you and possibly @sven-codeculture are the only ones reporting it and we do not know yet if it is the same thing or not.

I am glad to help, but I do not have the resources to investigate a bug I cannot reproduce myself, sorry for that.

**Comment by adamreisnz** [comment]
In my 20 years of experience with both closed and open source software, I have found that if one person reports an issue, there'll be at least another 99 people experiencing the issue but choosing not to report it (perhaps they are too busy, perhaps they are too lazy, perhaps they just don't care enough to report it and choose to switch to another software instead).

Personally, if this were my software and a bug like this were reported, I would be more worried and spend some time investigating it with real-life cases, because it sounds like a potential serious problem and it could cost me paying customers (for taskforce.sh). 

I also think recommending people to downgrade to an older version doesn't seem like a very productive approach, as there are reasons we are now at a newer version (e.g. security fixes, other bug fixes). 

I also don't think the "cannot reproduce" label is warranted at this point, since you admitted that you don't have the resources to properly investigate this issue at this point, so I don't think it can be concluded that it can't be reproduced just yet.

Lastly, shouldn't this version of BullMQ carry the label "beta version" if it isn't feature complete yet?

Thanks for your time looking into this so far.

**Comment by manast** [comment]
I am not going to start an argument here. The point is, unless you can provide a reproducible test case that reproduces the bug then it is not reproducible, and hence the label.

**Comment by sven-codeculture** [comment]
I've added some more logging to my job that was always hanging. Furthermore i've checked the whole code and noticed that there was a "await" missing at one position. It meight be possible that this was caused by the missing await.

Here is what we do (yea, we still use plain js instead of typescript, didn't have the time to convert our code to ts yet and it was originally started in normal js):
```
module.exports = (router, queue) => async (job) => {
   //doSomeStuff
  queue.addJob('someSubJob', {subJob:data}); // At this position we had an await missing inside some if that was not called every time
}
```
Still not sure why the missing await caused the whole bull job to be stuck, but since we changed this it was not stuck a single time in the last 5 days.

**Comment by manast** [comment]
@sven-codeculture you are also missing an await (or return) in the code you pasted.

**Comment by sven-codeculture** [comment]
none of our jobs actually does a return unless something failed, in this case we do a return Promise.reject(new Error("Some message"));
There was never any problem with that.

**Comment by manast** [comment]
In the code above if you do not return the result of calling "addJob" your will miss that error (best case it will show up as an unhandled exception). I recommend you lint your code with something like https://www.npmjs.com/package/eslint-plugin-promise
that will prevent similar issues in the future.

**Comment by sven-codeculture** [comment]
this was only the code that caused seemingly some hanging for us.
every queue.add, job.remove, etc. is now using await after some refactoring of the code. We can't just return every await queue.addJob since most of them are running in loops.

**Comment by tboulis** [comment]
We got the same issue after updating from `1.37.1` to `1.39.3`

**Comment by manast** [comment]
@tboulis any more info? 

Also remember, just because you experience this issue when upgrading it does not mean it is a bug in BullMQ (even if it could be of course), but it could also be the result of a fix or code change that now exposes a bug in user code.

**Comment by manast** [comment]
I also will stress the fact that as long as the issue is not reproducible I will not spend time working on it.

**Comment by tboulis** [comment]
@manast I totally agree. We are trying to figure out if it is our code's fault. I'll get back to you as soon as we figure it out. Until then, you shouldn't worry about it. I just mentioned it to keep the thread alive

**Comment by manast** [comment]
Did anybody gain any new insights? Something I though you could do when this happens is to check if there is a "lock" key in redis for the jobs that are active, so say you have jobId 1234 in a queue called "test", that has been stuck in active state for a long time, check if there is a key in redis like this: ```bull:test:1234:lock```  If this key is there then it means that the worker is still updating the lock key at least every 30 seconds (the default lock duration), so the worker believes the job has not completed yet, in this case I would most confidently believe the bug is in user code. If the "lock" key is not there, then it means the worker has stopped updating the lock for some reason, here there could be many reasons but if your QueueScheduler instance is working normally then these jobs that have lost their lock should be marked as stalled and moved back to the "wait" list. Since this is not happening it means the QueueScheduler is either missing or lost connection to redis.

**Comment by david-sykora** [comment]
Hello,

maybe I can offer a reproducible test case, which I isolated when we found similar symptoms in my project. I don't know if it is relatable to this issue, but the described behavior is similar. Also maybe it's more misuse than a bug.

steps to reproduce:
1. window `node ./lib/firstWorker.js`
2. window `node ./lib/secondWorker.js`
3. window `node ./lib/jobCreator.js`
4. window `node ./lib/jobCreator.js`

the problem of the code is, that the my `FancyQueueJob` wrapper around the job is creating `QueueEvents` instance for every job processed. In my environment the application stucks after few processed jobs because the `firstWorker` has active jobs, which are not really active and they are missing locks. Probably its fault of closed/failed redis connection, which are not able to live next to each other.

Reusing of single `QueueEvents` will solve that case.

Hope that this will help to someone :)

[demo.tar.gz](https://github.com/taskforcesh/bullmq/files/6956309/demo.tar.gz)

**Comment by adamreisnz** [comment]
After running fine for a couple of weeks, we now have run into the issue again, and the tasks are once again stuck in the active queue, despite Taskforce.sh reporting them as "failed":

![image](https://user-images.githubusercontent.com/490562/129460500-4c5e1192-0626-4a5a-b46d-a03f791c0416.png)

![image](https://user-images.githubusercontent.com/490562/129460491-54b20640-aa6c-409f-9b7b-3eb653aad17a.png)

@manast how come these are not moved to "Failed"?
They have had 5 failed attempts, and that is the configured max limit for each task.

Why are the tasks still in the active queue?

Clearly if they are reported as failed, it's not an issue of a "silent failure" somewhere outside of the async/await chain.
It is being picked up by BullMQ but not being acted on properly.

As a result, the queue has a backlog of 800 items to process which are not being processed now.
This is making BullMQ very unreliable for us.

What compounds the problem, is that removing the task via the Taskforce.sh interface doesn't work. The job is still there after a refresh of the page. The only thing that works is obliterating the queues, causing the pending tasks to get lost.

I will leave it in this state for a while in case you want to inspect the task.

**Comment by manast** [comment]
@adamreisnz 

>  and the tasks are once again stuck in the active queue, despite Taskforce.sh reporting them as "failed"

They are not reported as "failed". "Failed reason" is the failed reason for the last time this job failed, not that the job is in status failed as you seem to deduce. The job is still in status "active".

> @manast how come these are not moved to "Failed"?
They have had 5 failed attempts, and that is the configured max limit for each task.

Because the job does not seem to have been completed, if this is the fifth retry attempt, once the job fails it will not be retried. But it does not seem to be the case since it is still active.

> So why are the other 2 tasks still in the active queue?

Probably because the workers are still working on them. Put logs so that you can see how the workers are behaving and maybe you can figure out where they get stuck.

> Clearly if they are reported as failed, it's not an issue of a "silent failure" somewhere outside of the async/await chain.
It is being picked up by BullMQ but not being acted on properly.

They were reported as failed on the previous attempts, but it seems the jobs do not finalize on this last attempt, so this does not discard the possibility that the worker is for some reason never completing or failing.

> What compounds the problem, is that removing the task via the Taskforce.sh interface doesn't work. The job is still there after a refresh of the page. The only thing that works is obliterating the queues, causing the pending tasks to get lost.

Thats because jobs that are locked by a worker fail when they are removed. This is actually an issue with Taskforce that should not allow to remove active jobs, but other than that BullMQ is working as designed.

Unfortunatelly I do not see in your description anything that clearly would suggest a bug in BullMQ. As mentioned I recommend that you put more logs in your workers so that you can find out where they get stuck or what is going on in this situation where the job stays in active state forever. Just ask yourself this question: are your workers completing or failing in this situation where they stay as active? can you demonstrate for example that the job has completed/failed but due to some bug in BullMQ the job stays as active?  For now it seems that you do not really know if the worker has really finalized so I think thats where you should put your investigation efforts.

**Comment by adamreisnz** [comment]
@manast can you clarify, if the task says "attempts: 5", does that not mean it has already done 5 attempts?
That is the limit, so to me it seems like it's stuck after 5 attempts and not being moved to the failed queue.

Do you not think it's a coincidence that all tasks that are stuck get stuck at 5 attempts?

If this was a script error, you would expect it to get stuck at the 1st attempt wouldn't you?
Or at least see some variation.

Our script errors and exits correctly, I have tested and verified this, and this is not an issue with the script itself as far as we could investigate and deduce. There is a clear error at the end of it which is capture and processed and causes the task to fail as expected.

This is why it moves from the first failure (attempt 1) to attempt 2 and so forth, all the way to attempt 5, which is the limit, after which it gets stuck.

There is nothing special that happens on our end on the 5th attempt that would cause it to get stuck by itself.

Look I'm not trying to be annoying here, I'm just trying to understand and based on what you wrote, I still don't have a clear explanation for why this task would get stuck only after the attempt limit is reached, while our script is errorring properly.

**Comment by manast** [comment]
> Instead, it's consistent and repeatable. The task always gets stuck on the last attempt.

Then you should be able to present a test case that is reproducible. I took some time to write one that works as it should:

```ts
const IORedis = require("ioredis");
const { Queue, QueueScheduler, Worker } = require("bullmq");

const main = async () => {
  const qs = new QueueScheduler("Paint", {
    connection: new IORedis(),
  });
  const queue = new Queue("Paint", {
    connection: new IORedis(),
  });

  const worker = new Worker(
    "Paint",
    async (job) => {
      console.log(
        `worker: ${job.id} ${job.queueName} ${job.data.foo} ${job.opts.attempts}`
      );
      throw new Error();
    },
    {
      connection: new IORedis(),
    }
  );

  worker.on("error", (err) => {
    console.log(err);
  });

  worker.on("failed", async (job) => {
    console.log(`failed ${job.id}`);
  });

  worker.on("completed", (job, result) => {
    console.log(`completed ${job.id}`, result.slice(0));
  });

  const addJob = async (q) => {
    const j = await q.add(
      "car",
      { foo: "bar" },
      { attempts: 5, backoff: { type: "fixed", delay: 1000 } }
    );

    console.log(`added job ${j.id} on queue ${j.queueName}`);
  };

  let id = 0;
  addJob(queue);
};

main();
```

Gives:

```
added job 1 on queue Paint
worker: 1 Paint bar 5
failed 1
worker: 1 Paint bar 5
failed 1
worker: 1 Paint bar 5
failed 1
worker: 1 Paint bar 5
failed 1
worker: 1 Paint bar 5
failed 1
```

Examining with redis-cli:
```
$ redis-cli
keys *
1) "bull:Paint:events"
2) "bull:Paint:1"
3) "bull:Paint:id"
4) "bull:Paint:failed"
5) "bull:Paint:delay"
6) "bull:Paint:meta"
127.0.0.1:6379> zrange bull:Paint:failed 0 -1
1) "1"
```

**Comment by adamreisnz** [comment]
We have already done this as I wrote earlier.
The logs indicated the task has ended or errored. Nothing to indicate that
the script in user land is locked or stuck.

There was also a lock file present, but the worker was most definitely not
updating this every 30
seconds, because even shutting down the workers and restarting them didn't
jump start the queue.

Only a flush worked, as experienced by the case above as well. Seems like
the same symptoms.

**Comment by adamreisnz** [comment]
Having TTL would really help here....

**Comment by manast** [comment]
> There was also a lock file present, but the worker was most definitely not updating this every 30

That is impossible unless there is a bug in Redis since the lock has a 30 seconds expiration, if the worker was not updating it it would expire.

The code that runs after a job has completed is really simple. We have written code that shows it works. You mentioned it happens *always* so I expect you can reproduce this issue otherwise there is nothing else we can do here.

**Comment by adamreisnz** [comment]
Look I'm not making this stuff up, and I am just reporting what I am observing and seeing.
We have 75,000 users relying on our software working _properly_ and _reliably_. Trust, me, when the queue got stuck for days, I tried everything. And Heroku automatically restarts **all** dynos every day, so it most certainly is not impossible as you put it. I have seen it first hand.

The queue was stuck for literally days, while the workers would have been restarted several times in that timespan.

I am telling you, this really does not look like a userland error. Whether it's Redis or BullMQ, I don't know, but I do know it's a _big_ reliability problem. Judging by @pooreboy's notes that everything works fine when he migrates back to older version of Bull, seems to me the issue lies in the new BullMQ version somehow.

The only thing that works is nuking the entire queue.

**Comment by manast** [comment]
@adamreisnz you are spending a lot of time writing stuff that does not help in resolving this issue unfortunatelly. You can do a lot more investigation from your part, I have not access to your specific codebase how can I blindly discover what the problem is? you on the other hand have access to the whole BullMQ source code so you can also take a look to see if you see any problem...

**Comment by adamreisnz** [comment]
So do you, to be honest, but the difference is that you are the maintainer
/ creator of this package, and you are also trying to monetise it in the
form of Taskforce.sh, so I would have expected you to take this issue more
seriously and actually try and help resolve it instead of blaming it on
your end users.

There is nothing to share in terms of codebase or examples. The problems
don't happen with a simple contrived example. They happen in the production
environment under real world loads, with real world infrastructure.

So far, every reply of yours either blamed this on userland or explained
how what we are describing as a problem is "impossible". It's hard not to
get frustrated from that.

**Comment by manast** [comment]
@adamreisnz I recommend you that you re-read this thread again and check if I have not tried to help you and gave you hints on what to look after, but really as long as it cannot be reproduced what else do you expect me to do?

Now a piece of friendly warning: I honestly do not have time nor want to spend energy in discussing with you what are my obligations regarding the maintenance of this project, so next comment that is unrelated to the actual facts of this issue and does not help in finding a solution will end with closed issue and locked from further comments.

**Comment by duylddev** [comment]
I've experienced that for a week and decided to report it here.
1. I don't want to back to the old version because of addBulk feature (I add 3-5 million jobs once). If anybody knows how to do this trick on the old version, please tell me
2. I have 6-10 workers. Sometimes they restart cause of many reasons => jobs are stuck in active tab. In the old Bull version, after TTL jobs will be moved to failed, and I easily click retry all button 

=> Worker's stopped is the reason,

**Comment by manast** [comment]
@duylddev Do you have a QueueScheduler running as well? In BullMQ, the Scheduler must be explicitly instantiated to move the stalled jobs back to "wait" status.

**Comment by duylddev** [comment]
No, I haven't
Sorry about that. I'm adding it now
Thank you for your help

**Comment by adamreisnz** [comment]
I've pushed this update through right now, because our mail queue did get stuck again despite my previous fix.
It appears the same symptoms as before.

Two locks appear to be in place and these are holding everything up:

![image](https://user-images.githubusercontent.com/490562/140232771-b665bad6-b3d4-4371-8332-0a4aa7b24db6.png)

How long does it normally take for these locks to expire? These seem to have been stuck now for about 8 hours.

The TTL on the lock record is 24. Does that mean 24 hours?

![image](https://user-images.githubusercontent.com/490562/140232997-3897f3b7-033d-4f9c-aeb2-39db1e847c75.png)

After I manually removed the locks, the queue started operating again.

Looking at the logs of the job runner, the last 2 jobs that ran in that queue failed on account of an error with PDF generation, but given that the error was logged, the error was definitely captured cleanly and processed after job failure. However, I noted the run time of these jobs was significant, a few hours.

![image](https://user-images.githubusercontent.com/490562/140232946-fff730b1-3e76-4d79-98fa-e6b92364db2e.png)

Could there be an issue that when long running jobs don't complete in time, the bull mq queue somehow gets stuck?

We will obviously look into why those jobs were taking so long, but given that the logs did indicate the job (eventually) completed, it would be good to find out why the queue didn't progress afterwards.

If you have any further tips for what I can do to help analyse this that'd be great.

**Comment by adamreisnz** [comment]
Here is what bull MQ seems to have logged about that job:
![image](https://user-images.githubusercontent.com/490562/140233559-6fca61b0-9efd-4840-b867-ba3bd9896065.png)

**Comment by manast** [comment]
Ok, so just to highlight some facts that can help clarify what is going on: 
 - TTL Above is 24 seconds. So if that lock is never expiring then it is because a worker is renewing the lock.
 - Locks do not keep queues stopped. First of all, what is your concurrency factor and amount of workers? If there are two active locks it means there are 2 active jobs. If you have more than 2 workers or more than 2 concurrency factor in total, other workers will continue consuming jobs from the queue. When you manually delete the locks then the jobs that had the locks will fail because they will not be able to extend the locks and they will instead pick up new jobs. So it is not that the queue is stuck, is that the workers are busy still doing something. 
 - The message "Job error: HtmlPdf.create() connection lost." Where in the code is this error being produced? inside the processor function or on the "failed" event handler?
- Have you double checked that the job that failed with "job stalled more than allowable limit" is really one of the jobs that kept the lock alive forever? Can you paste a screenshot that shows that the jobid is the same as in the locks as well as all the options used?

> Could there be an issue that when long running jobs don't complete in time, the bull mq queue somehow gets stuck?

Not that anybody is aware of.

> We will obviously look into why those jobs were taking so long, but given that the logs did indicate the job (eventually) completed, it would be good to find out why the queue didn't progress afterwards.

But the question is, did the job really complete? When a job completes an atomic script is executed that among other things removes the lock, if it fails it will not be able to complete, so you will not get a global completion event for example.

So unfortunatelly the last version will not resolve this problem for you. All the symptoms seems to point into the direction of workers that are stuck doing something, i.e. not returning either with error or success.

**Comment by adamreisnz** [comment]
> So unfortunatelly the last version will not resolve this problem for you. All the symptoms seems to point into the direction of workers that are stuck doing something, i.e. not returning either with error or success.

Question regarding this last comment – what would happen if a process that is running workers is forcefully terminated, e.g. unclean shutdown? What happens to the tasks that were in progress?

Could this be a reason for those tasks potentially never completing?

**Comment by manast** [comment]
> Question regarding this last comment – what would happen if a process that is running workers is forcefully terminated, e.g. unclean shutdown? What happens to the tasks/jobs that were in progress?

The jobs will be active until the QueueScheduler detects they are stalled and moved back to wait... BullMQ is designed to handle this case, since it is pretty common. The queue will not be affected by it and as long as there are workers available they will keep consuming jobs... but the symptoms point that there are no workers available, they seem to be doing something and therefore you can see the locks being updated and not auto-expired which would be the case if there were no workers alive for those jobs.

**Comment by lukepolo** [comment]
I was able to get stuck in this loop 

![image](https://user-images.githubusercontent.com/2066668/142279827-7ed823cf-f819-452c-b39d-a1096b99c40b.png)

in fact i have 10 jobs stuck because we have a max queue of 10.  

We're currently using `bullmq@1.49.0` 

We use queue pause / resume quite a bit and also gracefully kill our queues as well. These jobs may not end during the grace period.

**Comment by manast** [comment]
@lukepolo this can happen for a variety of reasons, as pointed out in this thread already. You can make sure that your workers really have completed these jobs by putting a console.log precisely before returning, if the console.log is never printed then something is stuck in your own worker, otherwise it would be a bug in BullMQ (but so far nobody has been able to provide a piece of code that reproduces this issue).

**Comment by lukepolo** [comment]
thats fair, we ran into this cause our system crashed. And doesn't seem to be pulled in again. but I can see a case were the job may never finish. so ill report back to validate that concern.

**Comment by manast** [comment]
@lukepolo you are aware that you need to have a QueueScheduler that moves the stalled jobs back to wait right?

**Comment by lukepolo** [comment]
Correct! Were planning on upgrading to the pro version here today , pretty stoked about the group id's / rate limiting stuff , great work.

**Comment by guilhermebkel** [comment]
Hey guys! I had the same problem _(some job was getting stuck on active forever and got all the other jobs stuck)_. I gave a look in all comments above and tried to log my worker to understand if the job was never returning any result as the @manast suggested _(I added the logs on the events triggered by the worker, per example on "active", "failed", etc)_. Below is what I've got:

- That is a log of a customer's queue that was reported to be stuck. When I gave I look in the logs, I was able to understand that the job process function was never returning any result _(the last log was of it being active on 15:31, but it never got any result after a lot of hours until we restarted the worker)_:
![image](https://user-images.githubusercontent.com/42042433/144469392-4cef8b2f-f131-4587-a1c2-e0cc10ab60ee.png)

- So because of that the lock of this job was being renewed a lot of times and never getting expired on redis:
![image](https://user-images.githubusercontent.com/42042433/144469760-cbfce477-a791-4adc-a47a-22d23b5c130d.png)

Being minded about that, I added a manual timeout inside the job process in other to understand if this problem was being caused by myself _(below you guys can see an example of my code)_:

- Before setting up a manual timeout
```ts
const customJobHandler = async (job) => {
    try {
    	return await processJob(job.data)
    } catch (error) {
    	// I usually do something with the error below
    	return Promise.reject(error)
    }
}

const worker = new Worker("DeliverMessage", async job => await customJobHandler(job))
```

- After setting up a manual timeout
```ts
const customJobHandler = async (job) => {
    try {
    	return new Promise(async (resolve, reject) => {
    		try {
    			// 10 minutes timeout
    			const timeout = setTimeout(() => reject(new Error('TimeoutError')), 10 * 60 * 1000)

    			const result = await processJob(job.data)

    			clearTimeout(timeout)

    			resolve(result)
    		} catch (error) {
    			reject(error)
    		}
    	})
    } catch (error) {
    	// I usually do something with the error below
    	return Promise.reject(error)
    }
}

const worker = new Worker("DeliverMessage", async job => await customJobHandler(job))
```

Before setting up a timeout, I usually got the problem 2 times a day, and now after setting up the timeout, I've been 3 days without that problem. The really weird stuff about it is that I usually log all the errors I've got when processing jobs and I haven't got any TimeoutError 🤔

So being minded about it, it is only a coincidence that my job didn't get stuck in the last 3 days, or I was able to solve this problem indirectly _(maybe some of the code I have added has changed something internally in Node.js that helped not getting the job stuck? I do not really know)_

I'll keep looking for how many days my queues will run without being stuck and I will report any imore nformation during the following test days.

**Comment by LosTigeros** [comment]
Sometimes I also get jobs stuck in active.
![image](https://user-images.githubusercontent.com/6253928/144705772-aa034350-b36f-429f-8175-3c07f78de4a5.png)

I don't know the reason yet though.

**Comment by manast** [comment]
@guilhermebkel what you are basically saying is that the issue was in your code and not in bullmq since the timeout actually solved the issue for you.

**Comment by manast** [comment]
@LosTigeros the concept of "stuck" in the UI you are using is not aligned with what "stucked" means in BullMQ, in fact jobs cannot become "stucked", a job can be considered stuck only when it is moved from active to wait when the processor is not able to renew the lock on the job, which usually is because 1) the worker has died or 2 the worker is so CPU intensive that it does not have time to renew the lock in a timely manner.

**Comment by LosTigeros** [comment]
So what could be the reason that job stays as active for longer than it's timeout value?
I've checked it couple of times and job can return data before the timeout but after it errors with "Missing lock [...]".
Is it possible to adjust the interval for how long it locks the job so it has to renew it after same amount of time (or a little bit longer) as timeout value?
Should the job have it's own timer to interrupt and stop/fail the task instead of specifying the timeout value?

**Comment by manast** [comment]
@LosTigeros As explained in this thread, timeout is not implemented in BullMQ since it is not possible to cancel promises in javascript. So you can use your own timeout logic as @guilhermebkel wrote above.

As to why you get the missing lock error, that is quite difficult to know without code that reproduces the issue, but it could happen if your job is marked as stalled, moved back to wait but the original job is still processing and trying to complete a job that it no longer owns the lock for.

**Comment by Keith-Hon** [comment]
I have this issue when I control-c the consumer process. The task it's running just get stuck in active queue forever...

**Comment by Keith-Hon** [comment]
repo to reproduce
https://github.com/Keith-Hon/bullmq-demo

**Comment by JustDoItSascha** [comment]
I have the same problem right now after a worker has been ungracefully shutdown. The stalled jobs are not being picked up again and new jobs are waiting as well.

**Comment by ZeChArtiahSaher** [comment]
Happens to me as well exclusively with repeatable jobs

**Comment by manast** [comment]
@JustDoItSascha @ZeChArtiahSaher  do you have a QueueScheduler running in your system? this is the most common mistake, specially coming from Bull.

**Comment by maxpain** [comment]
The same problem

**Comment by manast** [comment]
As this thread is getting too long and and does not have any actionable item for us to do I will close it now.

This does not mean we want to ignore a potential issue, If you have an issue similar to this, please create a new issue with reproducible code and more context regarding your particular case and we will look into it.

**Comment by jiangdi0924** [comment]
same issue...

**Comment by manast** [comment]
@jiangdi0924 you may not be aware of it, but if you really want some issue to be resolved, just writing "same issue..." on an issue that has been closed due to lack of information and/or reproducibility is not going to help you out. Please create a new issue with as much information regarding *your* specific use case, and we will surely look into it.

**Comment by jiangdi0924** [comment]
> you may not be aware of it, but if you really want some issue to be resolved, just writing "same issue..." on an issue that has been closed due to lack of information and/or reproducibility is not going to help you out. Please create a new issue with as much information regarding _your_ specific use case, and we will surely look into it.

 * bullmq  version 1.86.10
 Thanks for reply so quick，in project  we have a lots of Worker and QueueScheduler.  The queue gets stuck, causing all tasks to not continue unless the node project is restarted. We're not quite sure what's causing the problem, and we'll continue to track down the details of the issue. I'd also like to ask if there's any correlation with memory space, as I've found that this doesn't happen when the local dev environment has more memory space.

**Comment by officert** [comment]
@manast is the `QueueScheduler` still required? The docs say it is not required from v2.0 onwards. I'm running into this same issue with jobs getting stuck in the active queue. I'm still reading through this thread and will try some of the suggestions in here, but the main one I've seen is about the `QueueScheduler`. Thank you

**Comment by manast** [comment]
@officert The ```QueueScheduler``` is not required, in fact it is not even possible to instantiate it since there is no such class anymore. The best is if you describe your particular case in a new issue. I am going to lock this one as there is no real benefit in making this thread longer since there is nothing actionable in this closed issue anymore.
